pyop2/codegen/rep2loopy.py

```diff
@@ -0,0 +1,781 @@
+        parameters.wrapper_arguments = rename_nodes(parameters.wrapper_arguments, replace)
+        if parameters.layer_start in replace:
+            parameters.layer_start = replace[parameters.layer_start]
+        if parameters.layer_end in replace:
+            parameters.layer_end = replace[parameters.layer_end]
+
+    # scheduling and loop nesting
+    deps = instruction_dependencies(instructions, mapper.initialisers)
+    within_inames = loop_nesting(instructions, deps, outer_inames, parameters.kernel_name)
+
+    # generate loopy
+    context = Bag()
+    context.parameters = parameters
+    context.within_inames = within_inames
+    context.conditions = []
+    context.index_ordering = []
+    context.instruction_dependencies = deps
+
+    statements = list(statement(insn, context) for insn in instructions)
+    statements = list(s for s in statements if not isinstance(s, DummyInstruction))
+
+    domains = list(parameters.domains.values())
+    if builder.single_cell:
+        new_domains = []
+        for d in domains:
+            if d.get_dim_name(isl.dim_type.set, 0) == builder._loop_index.name:
+                # n = start
+                new_domains.append(d.add_constraint(isl.Constraint.eq_from_names(d.space, {"n": 1, "start": -1})))
+            else:
+                new_domains.append(d)
+        domains = new_domains
+        if builder.extruded:
+            new_domains = []
+            for d in domains:
+                if d.get_dim_name(isl.dim_type.set, 0) == builder.layer_index.name:
+                    # layer = t1 - 1
+                    t1 = parameters.layer_end
+                    new_domains.append(d.add_constraint(isl.Constraint.eq_from_names(d.space, {"layer": 1, t1: -1, 1: 1})))
+                else:
+                    new_domains.append(d)
+        domains = new_domains
+
+    assumptions, = reduce(operator.and_,
+                          parameters.assumptions.values()).params().get_basic_sets()
+    options = loopy.Options(check_dep_resolution=True, ignore_boostable_into=True)
+
+    # sometimes masks are not used, but we still need to create the function arguments
+    for i, arg in enumerate(parameters.wrapper_arguments):
+        if parameters.kernel_data[i] is None:
+            arg = loopy.GlobalArg(arg.name, dtype=arg.dtype, shape=arg.shape)
+            parameters.kernel_data[i] = arg
+
+    if wrapper_name is None:
+        wrapper_name = "wrap_%s" % builder.kernel.name
+
+    wrapper = loopy.make_kernel(domains,
+                                statements,
+                                kernel_data=parameters.kernel_data,
+                                target=loopy.CTarget(),
+                                temporary_variables=parameters.temporaries,
+                                symbol_manglers=[symbol_mangler],
+                                options=options,
+                                assumptions=assumptions,
+                                lang_version=(2018, 2),
```

Comment:
Do we want `(2019, 1)` here now?

Reply:
@inducer / @kaushikcfd any thoughts?

Reply:
@inducer is more appropriate to answer this one.

Reply:
There's no `(2019, 1)` version, at least as yet. And until we need to break compatibility in some way, the language version will stay at `(2018, 2)`.

Reply:
OK, thanks.